### PR TITLE
Support resource control via systemd

### PR DESCRIPTION
**Pull request by Furisto**
This adds support for resource control using dbus and systemd. Currently supported are:
- Starting and removing transient units
- Resource control of cpu, cpusets and pids

Not yet supported is resource control with rootless containers. I have this working locally, but it still needs some more work and testing. This means that the following tasks are still open and will be followed up in another PR:
- Support for rootless containers
- Resource restrictions defined in the unified section
- Memory resource restrictions
- Integration tests

Thank you to @nimrodshn for his previous work on this, which is also included in this PR.

### CODE REVIEW

File: crates/libcgroups/src/systemd/dbus/client.rs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        pid: u32,
+        parent: &str,
+        unit_name: &str,
+    ) -> Result<()> {
+        // To view and introspect the methods under the 'org.freedesktop.systemd1' destination
+        // and object path under it use the following command:
+        // `gdbus introspect --system --dest org.freedesktop.systemd1 --object-path /org/freedesktop/systemd1`
+        let proxy = self.create_proxy();
+
+        // To align with runc, yuoki will always add the following properties to its container units:
```
Comment (utam0k) [inline crates/libcgroups/src/systemd/dbus/client.rs:43]:
```suggestion
        // To align with runc, youki will always add the following properties to its container units:
```

**Comment by utam0k** [comment]
Please give me some more time to understand this PR.

File: crates/libcgroups/src/systemd/cpu.rs [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    fn is_realtime_requested(cpu: &LinuxCpu) -> bool {
+        if cpu.realtime_period().is_some() {
+            return true;
+        }
+
+        if cpu.realtime_runtime().is_some() {
+            return true;
+        }
+
+        false
```
Comment (utam0k) [inline crates/libcgroups/src/systemd/cpu.rs:70]: How about this?
```suggestion
        cpu.realtime_period().is_some() || cpu.realtime_runtime().is_some()
```

File: crates/libcgroups/src/systemd/cpuset.rs [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+impl CpuSet {
+    fn apply(
+        cpu: &LinuxCpu,
+        systemd_version: u32,
+        properties: &mut HashMap<String, Box<dyn RefArg>>,
+    ) -> Result<()> {
+        if systemd_version < 244 {
+            bail!(
+                "Systemd version ({}) is too old to support cpuset restrictions",
```
Comment (utam0k) [inline crates/libcgroups/src/systemd/cpuset.rs:38]:
```suggestion
                "systemd version ({}) is too old to support cpuset restrictions",
```

**Comment by utam0k** [comment]
Why don't you collect the property keys in `mod.rs` or something?
https://github.com/systemd/systemd/blob/20f7ada699f41882ed455bb6de4ab96a7c012427/src/core/dbus-cgroup.c#L453-L508

**Comment by Furisto** [comment]
@utam0k PTAL

**Comment by utam0k** [comment]
We'll continue here. @nimrodshn Thanks a lot.
https://github.com/containers/youki/pull/451

**Comment by Furisto** [comment]
@utam0k I think you closed the wrong PR. Did you want to close #120?

**Comment by utam0k** [comment]
> @utam0k I think you closed the wrong PR. Did you want to close #120?

oops! sorry!

**Comment by a1ien** [comment]
We can also remove systemd dependency from libcontainer crate since it not use this dependency at all.

**Comment by Furisto** [comment]
@a1ien Good catch! I forgot about that one.

**Review by utam0k — APPROVED** [review]
Awesome

**Pull request merged**